examples/eng_wasm_contracts/millionaires_problem_demo/src/lib.rs

```diff
@@ -0,0 +1,88 @@
+#![no_std]
+#![feature(proc_macro_gen)]
+#![feature(proc_macro_non_items)]
+
+
+extern crate eng_wasm;
+extern crate eng_wasm_derive;
+extern crate rustc_hex as hex;
+#[macro_use]
+extern crate serde_derive;
+extern crate serde;
+
+use eng_wasm::*;
+use eng_wasm_derive::pub_interface;
+use eng_wasm::String;
+use eng_wasm::from_utf8;
+use hex::ToHex;
+use hex::FromHex;
+use std::collections::HashMap;
+use serde::{Serialize, Deserialize};
+
+// State key name "millionaires" holding a vector of Millionaire structs
+static MILLIONAIRES: &str = "millionaires";
+
+// Struct representing a Millionaire
+#[derive(Serialize, Deserialize)]
+pub struct Millionaire {
+    address: H256, // field containing 32 byte hash type for millionaire's address
+    net_worth: U256, // field containing 32 byte uint for millionaire's net worth
+}
+
+// Public-facing secret contract function declarations
+#[pub_interface]
+pub trait ContractInterface{
+    fn construct();
+    fn add_millionaire(address: H256, net_worth: U256);
+    fn compute_richest() -> H256;
+}
+
+pub struct Contract;
+
+// Private functions accessible only by the secret contract
+impl Contract {
+    // Read secret contract state to obtain vector of Millionaires (or new vector if uninitialized)
+    fn get_millionaires() -> Vec<Millionaire> {
+        match read_state!(MILLIONAIRES) {
+            Some(vec) => vec,
+            None => Vec::new(),
+        }
+    }
+}
+
+impl ContractInterface for Contract {
+    // Empty constructor
+    #[no_mangle]
+    fn construct() {
+
+    }
+
+    // Add millionaire with 32-byte hash type for address and 32-byte uint for net worth
+    #[no_mangle]
+    fn add_millionaire(address: H256, net_worth: U256) {
+        // Read state to get vector of Millionaires
+        let mut millionaires = Self::get_millionaires();
+        // Append a new Millionaire struct to this vector
+        millionaires.push(Millionaire {
+            address: address,
+            net_worth: net_worth,
+        });
+        // Write the updated vector to contract's state
+        write_state!(MILLIONAIRES => millionaires);
+    }
+
+    // Compute the richest millionaire by returning the 32-byte hash type for the address
+    #[no_mangle]
+    fn compute_richest() -> H256 {
+        // Read state to get vector of Millionaires and obtain the struct corresponding to the
+        // richest millionaire by net worth
+        match Self::get_millionaires().iter().max_by_key(|m| m.net_worth) {
```

Comment:
If there are several equal maximums, only one is returned. This is Ok, if this was your intention. But maybe it is better to return a vector of addresses? I am not sure though.

Reply:
i felt returning multiple in this scenario gives you information about another person's net worth (in the case it is equal), which goes against the motivation of the millionaires' problem. It may be fine given this edge case, but that was my reasoning behind this, but don't personally care too much either way! Leaving it as is for now :)